### PR TITLE
go modules, go 1.16, golangci-lint, sdk 3.6.7

### PR DESCRIPTION
**Pull request by roobre**

### CODE REVIEW

**Comment by roobre** [comment]
Please keep in mind that this is an ambitious refactor, which involves changing the driver used to connect to oracle databases with a new one from the same authors. Ideally changes should not be breaking, but this _needs to be e2e tested_.

**Comment by roobre** [comment]
Snyk errors are related to missing files in the `vendor/` folder, and should be ignored.

**Comment by roobre** [comment]
I've manually ran integration tests against an Oracle database in RDS.

Compared with the previous release of the integration, this build produced the following output:

```diff
ubuntu@ip:~$ diff old.json new.json
4c4
<   "integration_version": "2.5.3",
---
>   "integration_version": "0.0.0",
30,39c30,39
<           "db.averageActiveSessions": 0.0578540666666667,
<           "db.backgroundCheckpointsPerSecond": 0.0166666666666667,
<           "db.backgroundCpuUsagePerSecond": 0.492851666666667,
<           "db.backgroundTimePerSecond": 0.0104233333333333,
<           "db.blockChangesPerSecond": 1.33333333333333,
<           "db.blockChangesPerTransaction": 20,
<           "db.blockChangesPerUserCall": 0.260869565217391,
<           "db.blockGetsPerSecond": 1.73333333333333,
<           "db.blockGetsPerTransaction": 26,
<           "db.blockGetsPerUserCall": 0.332298136645963,
---
>           "db.averageActiveSessions": 5.00666666666667e-05,
>           "db.backgroundCheckpointsPerSecond": 0,
>           "db.backgroundCpuUsagePerSecond": 0.559915,
>           "db.backgroundTimePerSecond": 0.00966696666666667,
>           "db.blockChangesPerSecond": 0,
>           "db.blockChangesPerTransaction": 0,
>           "db.blockChangesPerUserCall": 0.0869565217391304,
>           "db.blockGetsPerSecond": 0,
>           "db.blockGetsPerTransaction": 0,
>           "db.blockGetsPerUserCall": 0.0869565217391304,
44,48c44,48
<           "db.consistentReadGetsPerSecond": 73.6,
<           "db.consistentReadGetsPerTransaction": 1104,
<           "db.cpuTimeRatio": 65.5908947916079,
<           "db.cpuUsagePerSecond": 3.570335,
<           "db.cpuUsagePerTransaction": 214.2201,
---
>           "db.consistentReadGetsPerSecond": 1.2,
>           "db.consistentReadGetsPerTransaction": 18,
>           "db.cpuTimeRatio": 146.604527296937,
>           "db.cpuUsagePerSecond": 0.0577083333333333,
>           "db.cpuUsagePerTransaction": 3.4625,
55,57c55,57
<           "db.cursorCacheHitsPerAttempts": 73.8326848249027,
<           "db.databaseCpuTimePerSecond": 5.78540666666667,
<           "db.dbwrCheckpointsPerSecond": 0.0166666666666667,
---
>           "db.cursorCacheHitsPerAttempts": 83.0188679245283,
>           "db.databaseCpuTimePerSecond": 0.00500666666666667,
>           "db.dbwrCheckpointsPerSecond": 0,
60,61c60,61
<           "db.enqueueRequestsPerSecond": 15.2166666666667,
<           "db.enqueueRequestsPerTransaction": 913,
---
>           "db.enqueueRequestsPerSecond": 9.71666666666667,
>           "db.enqueueRequestsPerTransaction": 583,
66,68c66,68
<           "db.executionsPerSecond": 17.5333333333333,
<           "db.executionsPerTransaction": 263,
<           "db.executionsPerUserCall": 3.1583850931677,
---
>           "db.executionsPerSecond": 0.133333333333333,
>           "db.executionsPerTransaction": 2,
>           "db.executionsPerUserCall": 1.26086956521739,
73,76c73,76
<           "db.hardParseCountPerSecond": 0.0666666666666667,
<           "db.hardParseCountPerTransaction": 4,
<           "db.hostCpuUsagePerSecond": 5.33333333333333,
<           "db.hostCpuUtilization": 2.68366320026837,
---
>           "db.hardParseCountPerSecond": 0,
>           "db.hardParseCountPerTransaction": 0,
>           "db.hostCpuUsagePerSecond": 0.933333333333333,
>           "db.hostCpuUtilization": 0.469641060046964,
79,83c79,83
<           "db.libraryCacheHitRatio": 99.4140625,
<           "db.libraryCacheMissRatio": 0.730994152046784,
<           "db.logicalReadsPerSecond": 75.3333333333333,
<           "db.logicalReadsPerTransaction": 1130,
<           "db.logonsPerTransaction": 5,
---
>           "db.libraryCacheHitRatio": 100,
>           "db.libraryCacheMissRatio": 0,
>           "db.logicalReadsPerSecond": 1.2,
>           "db.logicalReadsPerTransaction": 18,
>           "db.logonsPerTransaction": 0,
86,88c86,88
<           "db.openCursorsPerSecond": 17.2,
<           "db.openCursorsPerTransaction": 1032,
<           "db.osLoad": 0,
---
>           "db.openCursorsPerSecond": 0.883333333333333,
>           "db.openCursorsPerTransaction": 53,
>           "db.osLoad": 0.349609375,
93,97c93,97
<           "db.recursiveCallsPerSecond": 155.583333333333,
<           "db.recursiveCallsPerTransaction": 9335,
<           "db.redoWritesPerSecond": 0.4,
<           "db.redoWritesPerTransaction": 6,
<           "db.responseTimePerTransaction": 333.7184,
---
>           "db.recursiveCallsPerSecond": 14.7333333333333,
>           "db.recursiveCallsPerTransaction": 884,
>           "db.redoWritesPerSecond": 0,
>           "db.redoWritesPerTransaction": 0,
>           "db.responseTimePerTransaction": 2.4594,
100c100
<           "db.rowsPerSort": 8.61776061776062,
---
>           "db.rowsPerSort": 2.95294117647059,
103,106c103,106
<           "db.sharedPoolFreePercentage": 24.5135545730591,
<           "db.softParseRatio": 99.6268656716418,
<           "db.sortsPerUserCall": 5.6304347826087,
<           "db.sqlServiceResponseTime": 0.034557150253702,
---
>           "db.sharedPoolFreePercentage": 24.4921264194307,
>           "db.softParseRatio": 100,
>           "db.sortsPerUserCall": 1.84782608695652,
>           "db.sqlServiceResponseTime": 0.00264451612903226,
108,117c108,117
<           "db.tableScansPerUserCall": 0.409937888198758,
<           "db.totalIndexScansPerSecond": 30.2666666666667,
<           "db.totalIndexScansPerTransaction": 1816,
<           "db.totalParseCountPerSecond": 17.2,
<           "db.totalParseCountPerTransaction": 1032,
<           "db.totalTableScansPerSecond": 2.26666666666667,
<           "db.totalTableScansPerTransaction": 34,
<           "db.userCallsPerSecond": 5.13333333333333,
<           "db.userCallsPerTransaction": 77,
<           "db.userCallsRatio": 3.33436885161023,
---
>           "db.tableScansPerUserCall": 0.369565217391304,
>           "db.totalIndexScansPerSecond": 1.5,
>           "db.totalIndexScansPerTransaction": 90,
>           "db.totalParseCountPerSecond": 0.883333333333333,
>           "db.totalParseCountPerTransaction": 53,
>           "db.totalTableScansPerSecond": 0.2,
>           "db.totalTableScansPerTransaction": 3,
>           "db.userCallsPerSecond": 0.533333333333333,
>           "db.userCallsPerTransaction": 8,
>           "db.userCallsRatio": 4.94623655913978,
125c125
<           "db.waitTimeRatio": 35.80812445463,
---
>           "db.waitTimeRatio": 0,
129c129
<           "disk.logicalReadsPerUserCall": 14.2950310559006,
---
>           "disk.logicalReadsPerUserCall": 6.02173913043478,
132,138c132,138
<           "disk.physicalReadBytesPerSecond": 285815.466666667,
<           "disk.physicalReadIoRequestsPerSecond": 16.4666666666667,
<           "disk.physicalReadsPerSecond": 0.333333333333333,
<           "disk.physicalWriteBytesPerSecond": 24686.9333333333,
<           "disk.physicalWriteIoRequestsPerSecond": 0.25,
<           "disk.physicalWriteTotalIoRequestsPerSecond": 2.3,
<           "disk.physicalWritesPerSecond": 0.0833333333333333,
---
>           "disk.physicalReadBytesPerSecond": 58436.2666666667,
>           "disk.physicalReadIoRequestsPerSecond": 3.56666666666667,
>           "disk.physicalReadsPerSecond": 0,
>           "disk.physicalWriteBytesPerSecond": 14779.7333333333,
>           "disk.physicalWriteIoRequestsPerSecond": 0,
>           "disk.physicalWriteTotalIoRequestsPerSecond": 0.95,
>           "disk.physicalWritesPerSecond": 0,
155,157c155,157
<           "memory.pgaAllocatedInBytes": 336859136,
<           "memory.pgaFreeableInBytes": 39190528,
<           "memory.pgaInUseInBytes": 274119680,
---
>           "memory.pgaAllocatedInBytes": 336924672,
>           "memory.pgaFreeableInBytes": 35848192,
>           "memory.pgaInUseInBytes": 277569536,
160,165c160,165
<           "memory.redoGeneratedBytesPerSecond": 289.6,
<           "memory.redoGeneratedBytesPerTransaction": 4344,
<           "memory.sortsRatio": 100,
<           "network.ioMegabytesPerSecond": 0.3,
<           "network.ioRequestsPerSecond": 20.6,
<           "network.trafficBytePerSecond": 11589.1333333333,
---
>           "memory.redoGeneratedBytesPerSecond": 0,
>           "memory.redoGeneratedBytesPerTransaction": 0,
>           "memory.sortsRatio": 0,
>           "network.ioMegabytesPerSecond": 0.0666666666666667,
>           "network.ioRequestsPerSecond": 6.48333333333333,
>           "network.trafficBytePerSecond": 113.083333333333,
167,168c167,168
<           "query.physicalReadsPerTransaction": 5,
<           "query.physicalWritesPerTransaction": 17,
---
>           "query.physicalReadsPerTransaction": 0,
>           "query.physicalWritesPerTransaction": 0,
170c170
<           "redoLog.waits": 3105,
---
>           "redoLog.waits": 3268,
172c172
<           "rollbackSegments.gets": 7133,
---
>           "rollbackSegments.gets": 7584,
177c177
<           "sga.hitRatio": 0.9774455724283376,
---
>           "sga.hitRatio": 0.9784791352736678,
180c180
<           "sga.logBufferRedoEntries": 31415,
---
>           "sga.logBufferRedoEntries": 32137,
182,187c182,187
<           "sga.sharedPoolDictCacheMissRatio": 0.010372515030050346,
<           "sga.sharedPoolLibraryCacheHitRatio": 0.6253562877665032,
<           "sga.sharedPoolLibraryCacheReloadRatio": 0.002616703939143144,
<           "sga.sharedPoolLibraryCacheShareableMemoryPerStatementInBytes": 34931588,
<           "sga.sharedPoolLibraryCacheShareableMemoryPerUserInBytes": 7000,
<           "sga.ugaTotalMemoryInBytes": 55261792,
---
>           "sga.sharedPoolDictCacheMissRatio": 0.010326086428780202,
>           "sga.sharedPoolLibraryCacheHitRatio": 0.6380249903671492,
>           "sga.sharedPoolLibraryCacheReloadRatio": 0.0025663166620613427,
>           "sga.sharedPoolLibraryCacheShareableMemoryPerStatementInBytes": 35066535,
>           "sga.sharedPoolLibraryCacheShareableMemoryPerUserInBytes": 7750,
>           "sga.ugaTotalMemoryInBytes": 62770544,
189c189
<           "sorts.memoryInBytes": 31321
---
>           "sorts.memoryInBytes": 35850
2080c2080
<         "name": "TEMP",
---
>         "name": "RDSADMIN",
2096,2097c2096,2097
<           "displayName": "TEMP",
<           "entityName": "ora-tablespace:TEMP",
---
>           "displayName": "RDSADMIN",
>           "entityName": "ora-tablespace:RDSADMIN",
2102,2104c2102,2105
<           "tablespace.spaceConsumedInBytes": 0,
<           "tablespace.spaceReservedInBytes": 12509028,
<           "tablespace.spaceUsedPercentage": 0
---
>           "tablespace.offlineCDBDatafiles": 0,
>           "tablespace.spaceConsumedInBytes": 832,
>           "tablespace.spaceReservedInBytes": 12509012,
>           "tablespace.spaceUsedPercentage": 0.006651204747425296
2112c2113
<         "name": "RDSADMIN",
---
>         "name": "USERS",
2128,2129c2129,2130
<           "displayName": "RDSADMIN",
<           "entityName": "ora-tablespace:RDSADMIN",
---
>           "displayName": "USERS",
>           "entityName": "ora-tablespace:USERS",
2135c2136
<           "tablespace.spaceConsumedInBytes": 832,
---
>           "tablespace.spaceConsumedInBytes": 776,
2137c2138
<           "tablespace.spaceUsedPercentage": 0.006651204747425296
---
>           "tablespace.spaceUsedPercentage": 0.006203527504810132
2178c2179
<         "name": "USERS",
---
>         "name": "TEMP",
2194,2195c2195,2196
<           "displayName": "USERS",
<           "entityName": "ora-tablespace:USERS",
---
>           "displayName": "TEMP",
>           "entityName": "ora-tablespace:TEMP",
2200,2203c2201,2203
<           "tablespace.offlineCDBDatafiles": 0,
<           "tablespace.spaceConsumedInBytes": 776,
<           "tablespace.spaceReservedInBytes": 12509012,
<           "tablespace.spaceUsedPercentage": 0.006203527504810132
---
>           "tablespace.spaceConsumedInBytes": 0,
>           "tablespace.spaceReservedInBytes": 12509014,
>           "tablespace.spaceUsedPercentage": 0
```

File: build/Dockerfile [hunk 1]
```diff
@@ -1,6 +1,6 @@
-FROM golang:1.15.2-buster
+FROM golang:1.16-buster
 
-ARG GH_VERSION='1.4.0'
+ARG GH_VERSION='1.6.0'
```
Comment (davidbrota) [inline build/Dockerfile:3]:
```suggestion
ARG GH_VERSION='1.6.2'
```

**Review by davidbrota — CHANGES_REQUESTED** [review]
Remember to replace validate with golangci-lint here
https://github.com/newrelic/nri-oracledb/blob/master/.github/workflows/prerelease.yml#L64
otherwise the CI pipeline won't start

thanks for the testing!

**Comment by roobre** [comment]
> Remember to replace validate with golangci-lint here

Done, great catch @davidbrota!

**Review by davidbrota — APPROVED** [review]

**Pull request merged**